// @nftgo/gorarity/dist/cjs/models/token.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Token = void 0;
const token_identifier_1 = require("./token-identifier");
const token_metadata_1 = require("./token-metadata");
const token_standard_1 = require("./token-standard");
class Token {
    constructor(tokenIdentifier, tokenStandard, metadata) {
        this._tokenIdentifier = tokenIdentifier;
        this._tokenStandard = tokenStandard;
        this._metadata = metadata;
    }
    get tokenIdentifier() {
        return this._tokenIdentifier;
    }
    get tokenStandard() {
        return this._tokenStandard;
    }
    get metadata() {
        return this._metadata;
    }
    static fromErc721(contractAddress, tokenId, traits) {
        return new Token(new token_identifier_1.EVMContractTokenIdentifier(contractAddress, tokenId), token_standard_1.TokenStandard.ERC721, token_metadata_1.TokenMetadata.fromTokenTraits(traits));
    }
    hasTrait(traitName) {
        return this.metadata.traitExists(traitName);
    }
    traitCount() {
        return this._metadata.stringTraits.size;
    }
}
exports.Token = Token;
